src/shader_recompiler/ir/srt_gvn_table.h

```diff
@@ -0,0 +1,157 @@
+
+        if (inst->MayHaveSideEffects()) {
+            return NextValueNumber(IR::Value(inst));
+        }
+
+        u32 vn;
+
+        switch (inst->GetOpcode()) {
+        case IR::Opcode::Phi: {
+            // hack to get to parity with main
+            // Need to fix ssa_rewrite pass to remove certain phis
+            std::optional<IR::Value> source = TryRemoveTrivialPhi(inst);
+            if (!source) {
+                const auto pred = [](IR::Inst* inst) -> std::optional<IR::Inst*> {
+                    if (inst->GetOpcode() == IR::Opcode::GetUserData ||
+                        inst->GetOpcode() == IR::Opcode::CompositeConstructU32x2 ||
+                        inst->GetOpcode() == IR::Opcode::ReadConst) {
+                        return inst;
+                    }
+                    return std::nullopt;
+                };
+                source = IR::BreadthFirstSearch(inst, pred).transform([](auto inst) {
+                    return IR::Value{inst};
+                });
+                ASSERT(source);
+            }
+            vn = GetValueNumber(source.value());
+            value_numbers[IR::Value(inst)] = vn;
+            break;
+        }
+        case IR::Opcode::GetUserData:
+        case IR::Opcode::CompositeConstructU32x2:
+        case IR::Opcode::ReadConst: {
+            InstVector iv = MakeInstVector(inst);
+            if (auto it = iv_to_vn.find(iv); it != iv_to_vn.end()) {
+                vn = it->second;
+                value_numbers[IR::Value(inst)] = vn;
+            } else {
+                vn = NextValueNumber(IR::Value(inst));
+                iv_to_vn.emplace(std::move(iv), vn);
+            }
+            break;
+        }
+        default:
+            vn = NextValueNumber(IR::Value(inst));
+            break;
+        }
+
+        return vn;
+    }
+
+    u32 NextValueNumber(IR::Value v) {
+        u32 rv = next_num++;
+        value_numbers[v] = rv;
+        return rv;
+    }
+
+    ValueNumberTable value_numbers;
+    u32 next_num;
+
+    using InstVector = boost::container::small_vector<u32, 8>;
+
+    InstVector MakeInstVector(IR::Inst* inst) {
+        ASSERT(inst->GetOpcode() != IR::Opcode::Identity);
+        InstVector iv;
+        iv.reserve(2 + inst->NumArgs());
+        iv.push_back(static_cast<u32>(inst->GetOpcode()));
+        iv.push_back(inst->Flags<u32>());
+        for (auto i = 0; i < inst->NumArgs(); i++) {
+            iv.push_back(GetValueNumber(inst->Arg(i)));
+        }
+        return iv;
+    }
+
+    // Temp workaround for something like this:
+    // [0000555558a5baf8] %297   = Phi [ %24, {Block $1} ], [ %297, {Block $5} ] (uses: 4)
+    // [0000555558a4e038] %305   = CompositeConstructU32x2 %297, %296 (uses: 4)
+    // [0000555558a4e0a8] %306   = ReadConst %305, #0 (uses: 2)
+    // Should probably be fixed in ssa_rewrite
+    std::optional<IR::Value> TryRemoveTrivialPhi(IR::Inst* phi) {
```

Comment:
A better workaround might be to add a post-pass to ssa rewrite that removes those dead phis after the pass has run. At least then this file doesn't need to deal with it and we can fix it in faster way when IR gains ability to track users

Reply:
this is the reason the trivial phi is still there, right? (from the linked paper)
> _Moreover, if a φ function could be successfully replaced, other φ functions
using this replaced value might become trivial as well. For this reason, we apply
this simplification recursively on all of these users_

so we need to apply this on the users of the phi if TryRemoveTrivialPhi succeeds, and so on until we reach a fixed point?